### PR TITLE
Move NFD api to a separate go mod

### PR DESCRIPTION
**Pull request by ArangoGutierrez**
The idea behind this patch is to make it easier for third party apps to import NFD API.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -7,6 +7,7 @@ FROM ${BUILDER_IMAGE} as builder
 
 # Get (cache) deps in a separate layer
 COPY go.mod go.sum /go/node-feature-discovery/
+COPY api/ /go/node-feature-discovery/api/
```
Comment (marquiz) [inline Dockerfile:10]: Just the go.mod?

File: api/go.mod [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+require (
+	github.com/gogo/protobuf v1.3.2
+	github.com/stretchr/testify v1.8.4
+	golang.org/x/exp v0.0.0-20240222234643-814bf88cf225
+	k8s.io/api v0.29.2
+	k8s.io/apimachinery v0.29.2
+	k8s.io/client-go v0.29.2
+	k8s.io/klog/v2 v2.120.1
+	sigs.k8s.io/node-feature-discovery v0.15.1
```
Comment (marquiz) [inline api/go.mod:13]: We shouldn't depend on the main pkg -> some refactoring needed

**Review by marquiz — COMMENTED** [review]
Thanks @ArangoGutierrez. Maybe we should put it under `api/nfd/v1alpha1`?

File: Dockerfile [hunk 3]
```diff
@@ -10,11 +10,11 @@ COPY go.mod go.sum /go/node-feature-discovery/
 
 WORKDIR /go/node-feature-discovery
 
-RUN go mod download
-
 # Do actual build
 COPY . /go/node-feature-discovery
 
+RUN go mod download
+
```
Comment (marquiz) [inline Dockerfile:17]: Doing this after the COPY nullifies the whole point of pre-loading go modules (for caching). Can't you copy the go.{mod,sum} of the new module above, similarly to the main module?

File: go.mod [hunk 4]
```diff
@@ -20,22 +19,23 @@ require (
 	github.com/spf13/cobra v1.8.0
 	github.com/stretchr/testify v1.8.4
 	github.com/vektra/errors v0.0.0-20140903201135-c64d83aba85a
-	golang.org/x/exp v0.0.0-20240112132812-db7319d0e0e3
-	golang.org/x/net v0.20.0
+	golang.org/x/exp v0.0.0-20240222234643-814bf88cf225
```
Comment (marquiz) [inline go.mod:22]: If/when doing any updates on dependencies, please submit a separate PR.

File: go.mod
Comment (marquiz) [inline go.mod:1]: Should(n't) we add a replace directive to use the api/nfd module from the local copy?

File: go.mod (hunk 4 above)
Comment (ArangoGutierrez) [reply go.mod:22]: was not intentional, result of `go mod tidy`

File: go.mod
Comment (ArangoGutierrez) [reply go.mod:1]: look at the last line of replace

File: hack/generate.sh [hunk 5]
```diff
@@ -26,14 +26,12 @@ ${K8S_CODE_GENERATOR}/generate-groups.sh client,informer,lister \
     "nfd:v1alpha1" --output-base=. \
     --go-header-file hack/boilerplate.go.txt
 
-rm -rf pkg/generated
-
-mv sigs.k8s.io/node-feature-discovery/pkg/generated pkg/
+rm -rf api/v1alpha1/{clientset,informers,listers}
```
Comment (marquiz) [inline hack/generate.sh:29]: Shouldn't this be like

```suggestion
rm -rf api/generated
```

Comment (ArangoGutierrez) [reply hack/generate.sh:29]: Good catch!

**Comment by ArangoGutierrez** [comment]
After https://github.com/kubernetes-sigs/node-feature-discovery/pull/1605, we can restart the conversation here

**Comment by ArangoGutierrez** [comment]
@marquiz PTAL

**Comment by ArangoGutierrez** [comment]
@marquiz rebased PTAL

**Comment by ArangoGutierrez** [comment]
@vsoch  this helps towards the goal of serving your use case.
